File: C/Caching/FIFO.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

int pageFaults(int pages[], int n, int capacity)
{
    int s[MAX_SIZE] = {0};
    int indexes[MAX_SIZE] = {0};
    int page_faults = 0;
    int i, j;
    for (i = 0; i < n; i++)
    {
        if (s[pages[i]] == 0)
        {
            if (capacity > 0)
            {
                s[pages[i]] = 1;
                capacity--;
                page_faults++;
                indexes[pages[i]] = i;
            }
            else
            {
                int min_index = i;
                int min_page = -1;
                for (j = 0; j < MAX_SIZE; j++)
                {
                    if (s[j] == 1 && indexes[j] < min_index)
                    {
                        min_index = indexes[j];
                        min_page = j;
                    }
                }
                s[min_page] = 0;
                s[pages[i]] = 1;
                page_faults++;
                indexes[pages[i]] = i;
            }
        }
    }
    return page_faults;
}

int main()
{
    int n, capacity, i;
    int pages[MAX_SIZE];
    printf("Enter the length of the array: ");
    scanf("%d", &n);
    printf("Enter the elements of the array: ");
    for (i = 0; i < n; i++)
    {
        scanf("%d", &pages[i]);
    }
    printf("Enter the capacity: ");
    scanf("%d", &capacity);
    printf("Answer: %d\n", pageFaults(pages, n, capacity));
    return 0;
}

File: Java/Caching/FIFO.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class FIFO {
    static int pageFaults(int pages[], int n, int capacity) {
        HashSet<Integer> s = new HashSet<>(capacity);
        Queue<Integer> idx = new LinkedList<>();
        int page_faults = 0;
        for (int i = 0; i < n; i++) {
            if (s.size() < capacity) {
                if (!s.contains(pages[i])) {
                    s.add(pages[i]);
                    page_faults++;
                    idx.add(pages[i]);
                }
            } else {
                if (!s.contains(pages[i])) {
                    int val = idx.peek();
                    idx.poll();
                    s.remove(val);
                    s.add(pages[i]);
                    idx.add(pages[i]);
                    page_faults++;
                }
            }
        }

        return page_faults;
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the length of array: ");
        int n = sc.nextInt();
        int pages[] = new int[n];
        System.out.println("Enter the elements of array: ");
        for (int i = 0; i < n; i++) {
            pages[i] = sc.nextInt();
        }
        System.out.println("Enter the capacity: ");
        int capacity = sc.nextInt();
        System.out.println("Answer: ");
        System.out.println(pageFaults(pages, n, capacity));
    }
}
